File: Bit manipulation/2235. Add Two Integers.java
class Solution {
    /*
    Let's add 3 (binary 0011) and 5 (binary 0101):

    num1 ^ num2 gives us 0010, which represents the XOR result (sum without carry).
    (num1 & num2) << 1 gives us 0001 (carry, shifted by 1 position to the left).
    The recursive call now adds 0010 (sum without carry) and 0001 (carry), which gives us 0011, representing the final sum (binary 6).
    */
    public int sum(int num1, int num2) {
        if(num2 == 0){
            return num1;
        }
        return sum(num1 ^ num2, (num1 & num2) << 1);
    }
}

File: Bit manipulation/2235. Add Two Integers.cpp
class Solution
{
public:
    /*
    num1 -> sum
    num2 -> carry
    XOR  1 ^ 1 = 0
    0 ^ 0 = 0
    1 ^ 0 = 1
    -> equivalent to the adding without carry

    AND 1 ^ 1 -> 1 -> calculate the carry
    << 1 ->

    num2 == 0 -> carry = 0;
    return num 1, since num1 is the sum.
    */
    int sum(int num1, int num2)
    {
        int sum = num1;
        while (num2 != 0)
        {
            sum = num1 ^ num2;
            num2 = (num1 & num2) << 1;
            num1 = sum;
        }
        return sum;
    }
};